_posts/2020-01-17-strong-parameters-migration-guide.markdown

```diff
@@ -0,0 +1,97 @@
+
+That gem was supported by the Rails team until the release of Rails 5.0. After that we were forced to migrate to `Strong Parameters`. Once that happened we started to get unofficial forks of `protected_attributes` that support the latest version of Rails, like [protected_attributes_continued](https://github.com/westonganger/protected_attributes_continued).
+
+At this point we strongly recommend to fully migrate to `Strong Parameters`, since the available options to keep `Protected Attributes` alive, have very limited support and can encounter security issues.
+
+### Migration
+
+The migration consists in moving the Mass Assignment restrictions from the models to the controllers. That means removing the `attr_accessible` and `attr_protected` calls from your models and add a new method to your models' controllers to handle the parameters.
+
+In a simple example, this is how a model and controller look using `Protected Attributes`:
+
+```ruby
+# app/models/user.rb
+
+class User < ActiveRecord::Base
+  attr_accessible :first_name, :last_name, :email
+
+  # ...
+end
+```
+
+```ruby
+# app/controllers/users_controller.rb
+
+class UsersController < ApplicationController
+  def create
+    User.create(params[:user])
+  end
+
+  # ...
+end
+```
+
+This is how it should look after migrating to `Strong Parameters`:
+
+```ruby
+# app/models/user.rb
+
+class User < ActiveRecord::Base
+  # ...
+end
+```
+
+```ruby
+# app/controllers/users_controller.rb
+
+class UsersController < ApplicationController
+  def create
+    User.create(user_params)
+  end
+
+  # ...
+
+  private
+
+    def user_params
+      params.require(:user).permit(:first_name, :last_name, :email)
+    end
+end
+```
+
+Sometimes you'll find that the logic is more complex than that, but most of the time it's straightforward to move the parameters. The main issue is that it takes a lot of time when you have a big number of models. That's why we created [rails_upgrader](https://github.com/fastruby/rails_upgrader). This gem will help you speed up the migration process by automating some of it.
+
+Once you install the gem in your project you can run any of these commands in the console:
+
+- `rails_upgrader go` (attempt to upgrade your models and controllers in place)
+- `rails_upgrader dry-run` (write strong parameter migrations in the terminal)
+- `rails_upgrader dry-run --file` (write strong parameter migrations to `all_strong_params.rb` file)
+
+Another useful gem is [moderate_parameters](https://github.com/hintmedia/moderate_parameters). This gem can be handy to determine what data is originating from within the app and what is coming from the internet. I recommend you to take a look at it as well.
+
+### Conclusion
+
+Migrating to `Strong Parameters` can be a tedious process, but with the tools and resources that we presented here, it can be a lot easier.
```

Comment:
```suggestion
Migrating to `Strong Parameters` can be a tedious process but with the tools and resources that we presented here, it can be a lot easier.
```